src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -738,12 +738,9 @@ private void customLeaveToken(DetailAST ast, Deque<VariableDesc> variablesStack)
      */
     public static boolean shouldCheckIdentWithMethodRefParent(DetailAST identAst) {
         final DetailAST parent = identAst.getParent();
-        boolean result = true;
-        if (parent.getType() == TokenTypes.METHOD_REF) {
-            result = parent.getFirstChild() == identAst
+        return parent.getType() != TokenTypes.METHOD_REF
+                || parent.getFirstChild() == identAst
```

Comment:
At first, I also wanted to suppress this violation. Such code may be easy to read, but hard to debug.
Then I decided to see what would change if I agreed with IDEA and found that some parts of the code becomes uncovered.
And this is a good reason to adopt such style.

Reply:
>but hard to debug.

?? Separate lines for expression easy to put break point, ... Can you explain complexity ?

In old style we supposed to have more problems with coverage. As it know that functional style hide all branches. This update is not functional styled but still should not make more branches.
It might, but there are a lot of ways to write code to let coverage tool demand more testing. But it not a reason to have code like this.

Something is weird.

Reply:
I think there is bug/missing implementation in Jacoco.
What we have here (1):

```java
boolean flag = true;
if (condition1) {
  flag = condition2 && condition3;
}
```

This is semantically equivalent to (2)
```java
boolean flag = !condition1
  || condition2 && condition3;
```

For (2) Jacoco is able to detect a missing branch. For (1) it doesn't. I'm going to report this to the Jacoco maintainers. Maybe there is an option in Jacoco configuration.

Reply:
I would prefer to extract condition2 and condition3 to some descriptively named Boolean, then use it and condition1 in return statement as for (2).

Reply:
I think that corrected version is good enough

Reply:
I have more been preferring 1 if we are talking code that follows that strict pattern. I also see that as easier to debug since we can break on the assignment to know if it is being set true or not for the return.

I do also agree that splitting complex conditions out into its own method makes things easier, but I feel thats more for complex conditions and not simpler ones.

Reply:
I'm going to drop this changes for now (but keep new test cases).

Reply:
If parent type is METHOD, we should not calculate parent.getFirstChild() of next expression
But in new code we do.

Let's move resolving this items in separate PR.
I am ok to suppress them by Javadoc for now.

Reply:
> If parent type is METHOD, we should not calculate parent.getFirstChild() of next expression But in new code we do.

Why? New code:
```java
        return parent.getType() != TokenTypes.METHOD_REF
                || parent.getFirstChild() == identAst
```
If the first part (`parent.getType() != TokenTypes.METHOD_REF`) evaluates to `true`, then the second part (`parent.getFirstChild() == identAst`) will not be checked.

Same for the old code:
```java
  if (parent.getType() == TokenTypes.METHOD_REF) {
            result = parent.getFirstChild() == identAst
```
If the first part (`parent.getType() == TokenTypes.METHOD_REF`) evaluates to `false`, then the `if` body will be skipped.